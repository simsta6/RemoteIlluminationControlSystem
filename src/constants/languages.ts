export const LANGUAGES = [
    { 
        code: "en", 
        label: "English", 
    },
    { 
        code: "lt", 
        label: "Lietuvi≈≥", 
    }
];