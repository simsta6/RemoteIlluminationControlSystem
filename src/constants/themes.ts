import { Theme } from "@react-navigation/native";

export const KEY_FOR_STORAGE = "defaultScheme";

interface Colors {
    primary: string;
    background: string;
    card: string;
    text: string;
    disabledText: string;
    icon: string;
    border: string;
    notification: string;
    modal: string;
    button: string;
    disabledButton: string;
}

export const AppDarkColors: Colors = {
    primary: "rgba(133, 170, 164, 1)",
    background: "rgba(48, 48, 48, 1)",
    card: "rgba(33, 33, 33, 1)",
    text: "rgba(255, 228, 234, 1)", 
    disabledText: "rgba(186, 168, 172, 1)", 
    icon: "rgba(255, 228, 234, 1)", 
    border: "rgba(51, 70, 69, 1)",
    notification: "rgba(76, 96, 68, 1)",
    modal: "rgba(66, 66, 66, 1)",
    button: "rgba(33, 33, 33, 1)",
    disabledButton: "rgba(74, 74, 74, 1)",
};

export const AppLightColors: Colors = {
    primary: "rgba(122, 85, 91, 1)",
    background: "rgba(255, 235, 236, 1)",
    card: "rgba(229, 208, 237, 1)",
    text: "rgba(122, 85, 91, 1)", 
    disabledText: "rgba(87, 74, 76, 1)", 
    icon: "rgba(122, 85, 91, 1)", 
    border: "rgba(204, 185, 186, 1)",
    notification: "rgba(179, 159, 187, 1)",
    modal: "rgba(255, 247, 248, 1)",
    button: "rgba(229, 208, 237, 1)",
    disabledButton: "rgba(171, 163, 173, 1)",
};

export const AppDarkTheme: Theme = {
    dark: true,
    colors: AppDarkColors,
};

export const AppLightTheme: Theme = {
    dark: false,
    colors: AppLightColors
};