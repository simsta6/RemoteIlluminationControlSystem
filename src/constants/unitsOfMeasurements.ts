export const UNITS_OF_MEASUREMENTS = {
    TEMPERATURE: "Â°C",
    POWER: "W",
    VOLTAGE: "V",
    CURRENT: "A",
    LUX: "lx",
    "%": "%" 
};