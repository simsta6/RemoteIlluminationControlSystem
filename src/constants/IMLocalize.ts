import i18n, { LanguageDetectorAsyncModule } from "i18next";
import { initReactI18next } from "react-i18next";
import AsyncStorage from "@react-native-async-storage/async-storage";
import * as RNLocalize from "react-native-localize";
import { LANGUAGES } from "./languages";
import en from "../translations/en.json";
import lt from "../translations/lt.json";


const LANG_CODES = Object.keys(LANGUAGES.map(lang => lang.code));

const LANGUAGE_DETECTOR: LanguageDetectorAsyncModule = {
    type: "languageDetector",
    async: true,
    detect: (callback) => {
        AsyncStorage.getItem("user-language", (err, language) => {
            if (err || !language) {
                if (err) {
                    console.log("Error fetching Languages from asyncstorage ", err);
                } else {
                    console.log("No language is set, choosing English as fallback");
                }
                const findBestAvailableLanguage = RNLocalize.findBestAvailableLanguage(LANG_CODES);
  
                callback(findBestAvailableLanguage?.languageTag || "en");
                return;
            }
            callback(language);
        });
    },
    init: () => { 
        // later
    },
    cacheUserLanguage: (language: string) => {
        AsyncStorage.setItem("user-language", language);
    }
};

const resources = {
    en,
    lt,
} as const;

i18n
    .use(LANGUAGE_DETECTOR)
    .use(initReactI18next)
    .init({
        compatibilityJSON: "v3",
        lng: "en",
        react: {
            useSuspense: false
        },
        interpolation: {
            escapeValue: false
        },
        resources,
    });