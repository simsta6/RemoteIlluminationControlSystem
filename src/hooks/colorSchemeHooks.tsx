import { useTheme } from "@react-navigation/native";
import { AppDarkColors, AppLightColors } from "../constants/themes";

export const useAppColors = () => {
    const { dark } = useTheme();

    return {
        colors: dark ? AppDarkColors : AppLightColors
    };
};
