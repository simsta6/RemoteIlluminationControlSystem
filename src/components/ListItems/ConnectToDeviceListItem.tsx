import React from "react";
import { StyleSheet, Text, TouchableOpacity, View } from "react-native";
import { Device } from "react-native-ble-plx";
import { useAppColors } from "../../hooks/colorSchemeHooks";

interface Props {
    bleDevice: Device;
    connectOnPress: (bleDeviceId: string) => Promise<void>;
    isLast: boolean;
    isConnected: boolean;
}

export const ConnectToDeviceListItem = (props: Props) => {
    const { colors } = useAppColors();
    const { bleDevice, connectOnPress, isLast, isConnected } = props;

    return (
        <>
            <View style={styles.mainContainer}>
                <TouchableOpacity onPress={() => connectOnPress(bleDevice.id)}>
                    <View style={styles.verticalContainer}>
                        <View style={styles.container}>
                            <Text style={{ ...styles.deviceName, color: isConnected ? "green" : colors.text }}>{bleDevice.name ?? bleDevice.id}</Text>
                        </View>
                    </View>
                </TouchableOpacity>
            </View>
            {!isLast && <View style={{...styles.horizontalSeparator, backgroundColor: colors.text }}/>}
        </>
    );
};

const styles = StyleSheet.create({
    mainContainer: {
        marginVertical: 0
    },
    container: {
        flexDirection: "row",
        justifyContent: "space-between",
        alignItems: "center",
    },
    verticalContainer: {
        flexDirection: "column",
        justifyContent: "space-between",
        marginBottom: 10,
    },
    horizontalSeparator: {
        width: "100%",
        height: 1,
        opacity: 0.3,
        marginBottom: 10,
    },
    deviceName: {
        fontWeight: "500",
        fontSize: 16,
    }
});
