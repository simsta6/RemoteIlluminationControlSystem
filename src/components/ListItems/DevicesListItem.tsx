import React from "react";
import { StyleSheet, Text, View } from "react-native";
import ThreeVerticalDotsIcon from "../../assets/icons/ThreeVerticalDotsIcon";
import { useAppColors } from "../../hooks/colorSchemeHooks";
import { Device } from "../../state/devices/connectedDevicesTypes";
import { IconButton } from "../Buttons/IconButton";

interface Props {
    device: Device;
    Icon?: JSX.Element;
    isLast: boolean;
    setIsModalVisible: React.Dispatch<React.SetStateAction<boolean>>;
}

export const DevicesListItem = (props: Props) => {
    const { colors } = useAppColors();
    const { device, Icon, isLast, setIsModalVisible } = props;

    return (
        <>
            <View style={styles.row}>
                <View style={styles.closeItems}>
                    {Icon ? Icon : <></>}
                    <Text style={{ ...styles.deviceName, color: colors.text }}>{device.name}</Text>
                </View>
                <View style={{...styles.closeItems, ...styles.threeDotsContainer}}>
                    <IconButton 
                        Icon={() => <ThreeVerticalDotsIcon color={colors.text} height={25} width={25} />} 
                        onPress={() => setIsModalVisible(true)}
                    />
                </View>
            </View>
            {!isLast && <View style={{...styles.horizontalSeparator, backgroundColor: colors.text }}/>}
        </>
    );
};

const styles = StyleSheet.create({
    row: {
        flexDirection: "row",
        justifyContent: "space-between",
        alignItems: "center",
        height: 50,
    },
    closeItems: {
        flexDirection: "row",
        height: "100%",
        justifyContent: "flex-start",
        alignItems: "center",
    },
    horizontalSeparator: {
        width: "100%",
        height: 1,
        opacity: 0.3,
        marginVertical: 5,
    },
    deviceName: {
        fontWeight: "500",
        fontSize: 16,
        marginLeft: 16,
    },
    threeDotsContainer: {
        marginRight: 10,
    }
});
