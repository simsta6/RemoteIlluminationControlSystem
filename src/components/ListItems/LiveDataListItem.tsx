import React from "react";
import { StyleSheet, Text, View } from "react-native";
import { useAppColors } from "../../hooks/colorSchemeHooks";

interface Props {
    isLast: boolean;
    itemName: string;
    itemData: string;
    unitOfMeasurement?: string;
    rightSideIcon?: JSX.Element
}

export const LiveDataListItem = (props: Props) => {
    const { colors } = useAppColors();
    const { isLast, itemName, itemData, unitOfMeasurement, rightSideIcon } = props;

    return (
        <>
            <View style={styles.row}>
                <View style={styles.verticalContainer}>
                    <Text style={{ ...styles.itemName, color: colors.text }}>{itemName}</Text>
                    <Text style={{ ...styles.itemData, color: colors.text }}>{`${itemData} ${unitOfMeasurement ?? ""}`}</Text>
                </View>
                {rightSideIcon ? rightSideIcon : <></>}
            </View>
            {!isLast && <View style={{...styles.horizontalSeparator, backgroundColor: colors.text }}/>}
        </>
    );
};

const styles = StyleSheet.create({
    row: {
        flexDirection: "row",
        justifyContent: "space-between",
        alignItems: "center"
    },
    verticalContainer: {
        flexDirection: "column",
        justifyContent: "space-between",
    },
    horizontalSeparator: {
        width: "100%",
        height: 1,
        opacity: 0.3,
        marginVertical: 5,
    },
    itemName: {
        fontWeight: "500",
        fontSize: 16,
    },
    itemData: {
        fontWeight: "400",
        fontSize: 14,
    },
});
