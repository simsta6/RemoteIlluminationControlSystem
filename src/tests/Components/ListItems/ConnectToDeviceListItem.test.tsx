import "@testing-library/jest-dom";
import { fireEvent, render, waitFor } from "@testing-library/react-native";
import React from "react";
import { Device } from "react-native-ble-plx";
import { ConnectToDeviceListItem } from "../../../components/ListItems/ConnectToDeviceListItem";
import { Colors } from "../../../constants/themes";
import * as hooks from "../../../hooks/colorSchemeHooks";

describe("<ConnectToDeviceListItem />", () => {
    const colors = {
        text: "purple",
    };

    const bleDevice = { id: "bleDeviceId"} as Device;

    beforeAll(() => {
        jest.spyOn(hooks, "useAppColors").mockImplementation(() => ({ colors: colors as Colors }));
    });

    test("Default ConnectToDeviceListItem should match snapshot", () => {
        const mockConnectOnPress = jest.fn();

        const { toJSON } = render(
            <ConnectToDeviceListItem {...{bleDevice, connectOnPress: mockConnectOnPress, isLast: false, isConnected: false}} />
        );

        expect(toJSON()).toMatchSnapshot();
    });

    test("should generate horizontal separator if isLast = false", async () => {
        const mockConnectOnPress = jest.fn();

        const { queryAllByTestId } = render(
            <ConnectToDeviceListItem {...{bleDevice, connectOnPress: mockConnectOnPress, isLast: false, isConnected: false}} />
        );

        await waitFor(() => {
            expect(queryAllByTestId("horizontalSeparator")).toHaveLength(1);
        });
    });

    test("should not generate horizontal separator if isLast = true", async () => {
        const mockConnectOnPress = jest.fn();

        const { queryAllByTestId } = render(
            <ConnectToDeviceListItem {...{bleDevice, connectOnPress: mockConnectOnPress, isLast: true, isConnected: false}} />
        );

        await waitFor(() => {
            expect(queryAllByTestId("horizontalSeparator")).toHaveLength(0);
        });
    });

    test("should call connectOnPress when TouchableOpacity is pressed", async () => {
        const mockConnectOnPress = jest.fn();

        const { getByTestId } = render(
            <ConnectToDeviceListItem {...{bleDevice, connectOnPress: mockConnectOnPress, isLast: true, isConnected: false}} />
        );

        fireEvent.press(getByTestId("TouchableOpacity"));
        expect(mockConnectOnPress).toHaveBeenCalledTimes(1);
    });
});