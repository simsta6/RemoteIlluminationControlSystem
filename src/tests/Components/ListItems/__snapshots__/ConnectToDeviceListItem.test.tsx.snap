// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ConnectToDeviceListItem /> Default ConnectToDeviceListItem should match snapshot 1`] = `
Array [
  <View
    style={
      Object {
        "marginVertical": 0,
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="TouchableOpacity"
    >
      <View
        style={
          Object {
            "flexDirection": "column",
            "justifyContent": "space-between",
            "marginBottom": 10,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
            }
          }
        >
          <Text
            style={
              Object {
                "color": "purple",
                "fontSize": 16,
                "fontWeight": "500",
              }
            }
          >
            bleDeviceId
          </Text>
        </View>
      </View>
    </View>
  </View>,
  <View
    style={
      Object {
        "backgroundColor": "purple",
        "height": 1,
        "marginBottom": 10,
        "opacity": 0.3,
        "width": "100%",
      }
    }
    testID="horizontalSeparator"
  />,
]
`;
